tracing/trace.go

```diff
@@ -0,0 +1,385 @@
+func (tw *traceWrapper) send() {
+	tw.hcTrace.Send()
+}
+
+func (tw *traceWrapper) ses() (ses []stackEntry) {
+	sw := tw.root
+	for {
+		ses = append(ses, sw.se)
+		if sw.child == nil {
+			return
+		}
+		sw = sw.child
+	}
+}
+
+func (tw *traceWrapper) findGoroutine(in []stackEntry) []stackEntry {
+	twSes := tw.ses()
+	for t, set := range twSes {
+		if in[0].samePkgMethodStr(set) {
+			merged := append(twSes[0:t+1], in...)
+			for m := range merged[t+1:] {
+				merged[t+m].traceID = twSes[0].traceID
+			}
+			return merged
+		}
+	}
+	return nil
+}
+
+// spanWrapper keeps track of the contexts created for the different spans
+// and closes children as new spans arrive.
+type spanWrapper struct {
+	se     stackEntry
+	ctx    context.Context
+	trace  *traceWrapper
+	parent *spanWrapper
+	child  *spanWrapper
+	hcSpan Span
+}
+
+func newSpanWrapper(ctx context.Context, t *traceWrapper, hcs Span, se stackEntry) *spanWrapper {
+	newSpan := &spanWrapper{
+		ctx:    ctx,
+		trace:  t,
+		hcSpan: hcs,
+		se:     se,
+	}
+	se.addFields(hcs)
+	return newSpan
+}
+
+func newRootSpanWrapper(t *traceWrapper, stack []stackEntry) (root,
+	child *spanWrapper) {
+	root = newSpanWrapper(t.ctx, t, t.hcTrace.GetRootSpan(), stack[0])
+	if len(stack) > 1 {
+		child = root.stackToSpan(stack)
+	} else {
+		child = root
+	}
+	return
+}
+
+func (sw *spanWrapper) add(ns string, val interface{}) {
+	structToFields(sw.hcSpan.AddField, ns, val)
+}
+
+func (sw *spanWrapper) log(lvl int, msg string) {
+	if sw.child != nil {
+		sw.child.done()
+	}
+	_, sLog := sw.hcSpan.CreateChild(sw.ctx)
+	msgStruct := struct {
+		Lvl int
+		Msg string
+	}{lvl, msg}
+	// This value is unfortunately not exported from onet/log/lvl.go
+	if lvl > -17 {
+		structToFields(sLog.AddField, "log", msgStruct)
+	} else {
+		structToFields(sLog.AddField, "error", msgStruct)
+	}
+	sw.se.addFields(sLog)
+	sLog.Send()
+}
+
+func (sw *spanWrapper) done() {
+	if sw.child != nil {
+		sw.child.done()
+	}
+	sw.parent.child = nil
+	sw.hcSpan.Send()
+}
+
+func (sw *spanWrapper) updateSE(se stackEntry) {
+	sw.se = se
+	se.addFields(sw.hcSpan)
+}
+
+func (sw *spanWrapper) stackToSpan(stack []stackEntry) *spanWrapper {
+	if len(stack) == 1 {
+		sw.updateSE(stack[0])
+		return sw
+	}
+	if sw.child != nil {
+		if sw.child.se.traceID != stack[1].traceID ||
+			stack[0].line != sw.se.line {
+			sw.child.done()
+		}
+	}
+	if sw.child == nil {
+		sw.child = sw.createChild(stack[1])
+		sw.updateSE(stack[0])
+	}
+	return sw.child.stackToSpan(stack[1:])
+}
+
+func (sw *spanWrapper) createChild(se stackEntry) *spanWrapper {
+	ctx, newHCS := sw.hcSpan.CreateChild(sw.ctx)
+	newChild := newSpanWrapper(ctx, sw.trace, newHCS, se)
+	newChild.parent = sw
+	return newChild
+}
+
+type stackEntry struct {
+	line      int
+	path      string
+	file      string
+	pkgPath   string
+	method    string
+	params    string
+	traceID   string
+	createdBy bool
+}
+
+var stackFileLine = regexp.MustCompile(
+	"\\s*(.*):([0-9]*)( \\+0x[a-f0-9]*$)?\\s*")
```

Comment:
Might be clearer to use a raw string, i.e. `` `\s*(.*):([0-9]*)( \+0x[a-f0-9]*$)?\s*` ``

Reply:
Thanks - I only discovered `` ` ` `` halfway through and then forgot to change the others ;)